File: 每日一题/剑指 Offer 07. 重建二叉树.py
# https://leetcode-cn.com/problems/zhong-jian-er-cha-shu-lcof/
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        dic = {}
        for i in range(len(inorder)):
            dic[inorder[i]] = i

        def recur(root, left, right):
            if left > right:
                return
            node = TreeNode(preorder[root])
            i = dic[preorder[root]]
            node.left = recur(root+1, left, i-1)
            # preorder[root,left_root,---,right_root,---]
            # inorder[left|root|right]
            # right_root=root+1+len(left)=root+1+i-left
            node.right = recur(i-left+root+1, i+1, right)
            return node
        return recur(0, 0, len(inorder)-1)

File: ts-version/剑指 Offer 07. 重建二叉树.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}
function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
    if (!preorder.length) return null;

    const root = preorder[0];
    const index = inorder.indexOf(root);
    const left = inorder.slice(0, index);
    const right = inorder.slice(index + 1);

    const rootNode = new TreeNode(
        root,
        left.length > 0 ? buildTree(preorder.slice(1, index + 1), left) : null,
        right.length > 0 ? buildTree(preorder.slice(index + 1), right) : null
    );

    return rootNode;
}
